core/arch/arm/kernel/user_ta.c

```diff
@@ -657,6 +658,12 @@ static TEE_Result alloc_and_map_ldelf_fobj(struct user_ta_ctx *utc, size_t sz,
 	return res;
 }
 
+static void gdb_helper(const vaddr_t ldelf_addr, const TEE_UUID uuid)
+{
+	(void)ldelf_addr;
+	(void)uuid;
+}
```

Comment:
1. Does this help even with `DEBUG=0`?, in other words with `-Os` instead of `-O0`?
[QEMU](https://github.com/OP-TEE/build/blob/3.9.0/qemu.mk#L24) in https://github.com/OP-TEE/build is built by default with `DEBUG=1` but QEMUv8 isn't. And, some issues might happen only in non-debug mode such as https://github.com/OP-TEE/optee_os/pull/3934 (although not a very common thing).
2. Normally we add `__unused` to the parameters instead of doing the `(void)` cast. Would it work too?

Reply:
> * Does this help even with `DEBUG=0`?, in other words with `-Os` instead of `-O0`?
>   [QEMU](https://github.com/OP-TEE/build/blob/3.9.0/qemu.mk#L24) in https://github.com/OP-TEE/build is built by default with `DEBUG=1` but QEMUv8 isn't. And, some issues might happen only in non-debug mode such as #3934 (although not a very common thing).

`DEBUG=0` with (Os) and the symbols are gone.

Related but unrelated, I first tried with `CFG_CC_OPTIMIZE_FOR_SIZE=y` (i.e., `Os`) I get a not so helpful error message (it'd be nice to know **why**is must be **n**).


 ```
core/arch/arm/arm.mk:135: *** CFG_CC_OPTIMIZE_FOR_SIZE is set to 'y' (from command line) but its value must be 'n'.  Stop
```

> * Normally we add `__unused` to the parameters instead of doing the `(void)` cast. Would it work too?

That would probably work just as well.


> Minor: I think the prefix should be core, ldelf: rather than gdb: because its main purpose is to identify roughly which area of the code is modified ("does this commit touch this or this...?").

Sure

> Also some more description about how the helper functions help would be welcome, perhaps just mention that some GDB script can intercept the calls and record the parameters, something like that?

I can elaborate more in the commit message.

Reply:
> Related but unrelated, I first tried with CFG_CC_OPTIMIZE_FOR_SIZE=y (i.e., Os) I get a not so helpful error message (it'd be nice to know why it must be n).
> ```
> core/arch/arm/arm.mk:135: *** CFG_CC_OPTIMIZE_FOR_SIZE is set to 'y' (from command line) but its value must be 'n'.  Stop`
> ```

For that you need to check the Makefile :-/
The error is the consequence of using `$(call force ...)` but the reason why is usually found in the context / comments at the call site. Looking at https://github.com/OP-TEE/optee_os/blob/3.9.0/core/arch/arm/arm.mk#L120-L124, in this case `DEBUG=1 CFG_CC_OPTIMIZE_FOR_SIZE=y` causes a conflict and we chose to preserve the meaning of `DEBUG=1`.

Reply:
> `DEBUG=0` (with `Os`) and the symbols are gone.

That's annoying. Would `__attribute__((optimize("O0")))` on the helpers work?

Reply:
> > Related but unrelated, I first tried with CFG_CC_OPTIMIZE_FOR_SIZE=y (i.e., Os) I get a not so helpful error message (it'd be nice to know why it must be n).
> > ```
> > core/arch/arm/arm.mk:135: *** CFG_CC_OPTIMIZE_FOR_SIZE is set to 'y' (from command line) but its value must be 'n'.  Stop`
> > ```
> 
> For that you need to check the Makefile :-/

Of course, but that still a Sherlock Holmes task :) And the reason for this is "# For backwards compatibility" :)  Anyhow, that was a side track, let's not det derailed.

Reply:
> > `DEBUG=0` (with `Os`) and the symbols are gone.
> 
> That's annoying. Would `__attribute__((optimize("O0")))` on the helpers work?

It did! Thanks! (Only added it to one of the two functions when testing)
```
$ find ../optee_os/out/ -name "*.mk" | xargs grep -e "O0\|O1\|O2\|O3\|Os"                                                                                                                                                                                                     
../optee_os/out/arm/export-ta_arm32/mk/conf.mk:ta_arm32-platform-cflags := -mcpu=cortex-a15 -Os -g3 -fpic -mthumb -fno-short-enums -fno-common -mno-unaligned-access -mfloat-abi=hard -funsafe-math-optimizations -funwind-tables

$ find ../optee_os/ -name "*.elf" | xargs nm -a | grep gdb                                                                                                                                                                                                                    
0e126989 t gdb_helper
nm: ../optee_os/out/arm/ta/avb/023f8f1a-292a-432b-8fc4-de8471358067.stripped.elf: no symbols
nm: ../optee_os/out/arm/ta/trusted_keys/f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c.stripped.elf: no symbols
nm: ../optee_os/out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.stripped.elf: no symbols
```

Reply:
Hmm, it's still a bit fragile. I would have tried playing with some volatile assignment inside the function or even adding an assembly function. Maybe even this works:
``` C
static void gdb_helper(const volatile vaddr_t ldelf_addr __unused, const volatile TEE_UUID *uuid _unused)
{
}
```

Reply:
We are usually very Linux kernel inspired, they do not encourage volatile:
https://www.kernel.org/doc/html/latest/process/volatile-considered-harmful.html

Having said, in this case, it don't see a problem with using volatile. Why do you still see the attribute trick still as fragile?

Reply:
Making assumptions on how the compiler optimizes at a certain level of optimization seems like you have no clear contract to me.

The best would probably be to add a
``` C
void gdb_helper(const vaddr_t ldelf_addr __unused, const TEE_UUID *uuid __unused)
{
}
```
in a separate C file. That way the compiler may optimize as much as it wants, we still have the calling convention to fall back to so the register content is well defined.

Reply:
@jenswi-linaro that's assuming the function is not removed entirely due to LTO (Clang).

Reply:
On the flip side, this is meant mainly for our QEMU debugging and there it seems to behave as expected. So I could live with leaving it as it is now and fix it if and when we see any problems. If it would have been key functionality running all the time in end products, then I'd reason differently.

Reply:
LTO, do we have that enabled? Why? Do we gain anything by it?

Reply:
> LTO, do we have that enabled? 

Hmm... no :wink: I mistakenly thought it was enabled by default with recent builds of Clang, but luckily it's not so my remark is not very relevant.